File: hellojava/src/main/java/com/fathzer/oop/AbstractThing.java
package com.fathzer.oop;

public abstract class AbstractThing {
    private String name;

    protected AbstractThing(String name) {
        if (name==null) {
            throw new IllegalArgumentException("name");
        }
        this.name = name;
    }
    
    public abstract void doSomething();

    public String getName() {
        return name;
    }
}

File: hellogo/pkg/oop/abstract_thing.go
package hellogo

type AbstractThing struct {
    name string
}

func NewAbstractThing(name string) *AbstractThing {
    if name == "" {
        panic("name cannot be empty")
    }
    return &AbstractThing{name: name}
}

func (t *AbstractThing) GetName() string {
    return t.name
}

// DoSomething must be implemented by concrete types
func (t *AbstractThing) DoSomething() {
    panic("DoSomething must be implemented")
}
